mqtt.cpp

```diff
@@ -329,29 +358,33 @@ int OSMqtt::_init(void) {
 
 	if (mqtt_client) { mosquitto_destroy(mqtt_client); mqtt_client = NULL; };
 
-	mqtt_client = mosquitto_new("OS", true, NULL);
+	mqtt_client = mosquitto_new(_id, true, NULL);
 	if (mqtt_client == NULL) {
 		DEBUG_PRINTF("MQTT Init: Failed to initialise client\n");
 		return MQTT_ERROR;
 	}
 
-	mosquitto_connect_callback_set(mqtt_client, _mqtt_connection_cb);
-	mosquitto_disconnect_callback_set(mqtt_client, _mqtt_disconnection_cb);
-	mosquitto_log_callback_set(mqtt_client, _mqtt_log_cb);
-	mosquitto_will_set(mqtt_client, MQTT_AVAILABILITY_TOPIC, strlen(MQTT_OFFLINE_PAYLOAD), MQTT_OFFLINE_PAYLOAD, 0, true);
+	char availability_topic[MQTT_MAX_TOPIC_LEN + 1] = { 0 };
+	_make_full_topic(availability_topic, MQTT_AVAILABILITY_TOPIC);
+	mosquitto_will_set(mqtt_client, availability_topic, strlen(MQTT_OFFLINE_PAYLOAD), MQTT_OFFLINE_PAYLOAD, 0, true);
+
+	mosquitto_connect_callback_set(mqtt_client, _mqtt_callbacks::_mqtt_connection_cb);
+	mosquitto_disconnect_callback_set(mqtt_client, _mqtt_callbacks::_mqtt_disconnection_cb);
+	mosquitto_log_callback_set(mqtt_client, _mqtt_callbacks::_mqtt_log_cb);
 
 	return MQTT_SUCCESS;
 }
 
 int OSMqtt::_connect(void) {
-	int rc;
-	if (_username[0]) {
-		rc = mosquitto_username_pw_set(mqtt_client, _username, _password);
-		if (rc != MOSQ_ERR_SUCCESS) {
-			DEBUG_LOGF("MQTT Connect: Connection Failed (%s)\n", mosquitto_strerror(rc));
-			return MQTT_ERROR;
-		}
+
+	// Note: If (username == NULL) then disable authentication
+	//       If (username != NULL && password == NULL) then username is used without a password
```

Comment:
No, you cannot do that. See https://github.com/eclipse/mosquitto/blob/master/lib/options.c#L69
Please revert this change

Reply:
Thanks for the review. This is getting interesting as it looks like the latest version of mosquitto does not match the documentation! But we should probably reference this link [here](https://github.com/eclipse/mosquitto/blob/30c818ed615ad078db3760fee1efd21d84bf5e64/lib/mosquitto.c#L229) as this is the version 1.3.4 which is available by `apt` for OSPi (i.e. much earlier). But lets dig into this as the code isnt operating as I would expect.

The motivation behind this change was to ensure we deal with three use cases: 1) username not provided (un-authenticated), 2) both username and password are provided (authenticated); and 3) username is provided but password is not (unprotected user login).

It is this third case that I was considering as according to the mosquitto documentation and code I linked above then we should be calling `mosquitto_username_pw_set(mqtt_client, _username, NULL)` rather than passing an empty string as _password.

Appreciate your thoughts on this as it feels like a decision between coding for the currently distributed version vs what might come down the road.

Reply:
I had a wrong reading of source code. Apparently, it does not change anything but documentation says to implement it as you did.
https://mosquitto.org/api/files/mosquitto-h.html#mosquitto_username_pw_set
So, I agree with this change ;-)
Thanks

Reply:
I also pondered on this and so far I've assumed that if user name is non-empty then password is also non-empty. Would anyone intentionally set only a user name but empty password in practice?

Reply:
I agree, in normal practice you would want to use either case 1 or 2 but the spec does allow for case 3 as well. I can imagine scenarios in a trusted network where you want to track usage but aren't too worried about rogue agents ! 

Having said that, if we want to restrict against scenarion this then we should do it earlier in the process i.e. in the App UI or api response so we can provide user feedback.

Reply:
As documentation says to do it like this and you implemented it, let's merge this improvement